Add support for headstage-64 firmware version 0.4

jonnew opened the pull request:
- This PR supersedes #521 
- Fixes #502 
- Fixes #516 (supersedes #520)
- Fixes #537
- Fixes #542

jonnew committed:
Implement support for headstage-64 firmware 0.4

- Remove explcity ostim and estim trigger delay registers in favor of
  dynamic delay via trigger register
- Add the ability to turn off indication LED
- Fix register values and frame defintions
- Add persistant heartbeat device to Headstage64 device aggregate

jonnew committed:
Add Headstage64GpoTrigger

- This is a sink operator that takes booleans and can issue
  stimulus triggers to the headstage-64 optical and electrical
  stimulators using the port controller's GPO
- To implement this functionality, derivatives of ConfigurePortController
  now must explicitly specify their DeviceType. This allows the
  creation of specialized versions of the PortController for
  particular implementations of MultiDeviceFactory. In most cases,
  PortController will be used.
- In the case of ConfigureHeadstage64, a type called Headstate64PortController
  was created and is used by ConfigureHeadstage64's specialization of
  ConfigurePortController. This is also the device filter used by
  Headstage64GpoTrigger.
- This allows headstage-64 stimulus triggers using GPO, but will only work if a
  ConfigureHeadstage64 operator is in the configuration chain and does
  not permit general access to GPO since the meaning of each line depends on
  the headstage that is attached to the port.

bparks13 commented on OpenEphys.Onix1/ConfigureHeadstage64ElectricalStimulator.cs:
```diff
@@ -77,42 +73,20 @@ public ConfigureHeadstage64ElectricalStimulator(ConfigureHeadstage64ElectricalSt
         {
-            get => triggerDelay.Value;
-            set => triggerDelay.OnNext(value);
+            get => stimEnable.Value;
```
Question, should the backing field match the name of the property?

jonnew replied:
Since we have a pretty dedicated meaning for "Enable" across all the devices, I thought that renaming this would be easier to understand and that "Arm" was a good choice since the behavior of this register is actually just to make the sequencer responsive to triggers. There is no reason that a property must take the same name as a particular register, just like there is no need to expose every single register directly.

bparks13 commented on OpenEphys.Onix1/ConfigureHeadstage64OpticalStimulator.cs:
```diff
@@ -71,31 +70,31 @@ public ConfigureHeadstage64OpticalStimulator(ConfigureHeadstage64OpticalStimulat
         {
-            get => delay.Value;
-            set => delay.OnNext(Clamp(value, Headstage64OpticalStimulator.MinDelay, Headstage64OpticalStimulator.MaxDelay));
+            get => stimEnable.Value;
```
Same question as above

bparks13 commented on OpenEphys.Onix1/ConfigureHeadstageNeuropixelsV1e.cs:
```diff
@@ -2,6 +2,7 @@
 using System.Collections.Generic;
 using System.ComponentModel;
 using System.Threading;
+using static OpenEphys.Onix1.ConfigureHeadstage64;
```
Is this using needed?

bparks13 reviewed (commented):
Looks good, and is definitely good to restrict access to the GPOs unless the Headstage64 is present. I have one comment that needs to be addressed, and some small nitpicks, plus after this review I will go in and add the persistent heartbeat tab to the HS64 GUI for completeness.

bparks13 committed:
Add Heartbeat tab for Headstage64 dialog

jonnew commented on OpenEphys.Onix1/Headstage64GpoTrigger.cs:
```diff
@@ -0,0 +1,51 @@
+    /// trigger will be delivered to both of the stimulators so care must be taken to ensure only the
+    /// appropriate stimulator is armed when the trigger is sent. 
+    /// </remarks>
+    [Description("Controls a headstage-64 onboard optical stimulus sequencer.")]
```
good catch

jonnew committed:
Address code review

- Readd inter phase current property to
  ConfigureHeadstage64ElectricalStimulator
- Improve code comments

jonnew committed:
Remove Range attribute for ConfigurePersistentHeartbeat.BeatsPerSecond

- The range is a readonly register value that is device specific and not
  know at compile time

bparks13 reviewed (approved):
Looks good to me

cjsha reviewed (changes requested):
I locally merged this branch into main so I can test and document the stimulus reports that are already merged into main and the features of this branch with the hs-64 0.4 that we have. I couldn't get any Headstage64OpticalStimulatorDataFrames despite sending values to the corresponding trigger node. Also, I don't get a frame for the first estim trigger, but the rest report back just fine. I'm not sure where these behaviors are coming from (my workflow, software, firmware, etc.), but I don't want to let this slip through the cracks so I'm going to request changes here.

example workflow:
[hs64-example.zip](https://github.com/user-attachments/files/23243060/hs64-example.zip)

<img width="1920" height="1040" alt="image" src="https://github.com/user-attachments/assets/15b7e873-5dd9-4629-b6ca-8217375dd1c1" />

This spawned this issue: https://github.com/open-ephys/onix-common-hdl/issues/48

jonnew committed:
Allow Headstage64PortController to be used by PortStatus

jonnew commented:
This branch introduced #537 which is fixed by b84dcd2

cjsha reviewed (changes requested):
I got frames from `Headstage64OpticalStimulatorData`. However, some seem super delayed. Am I doing something wrong here?

<img width="1920" height="1040" alt="image" src="https://github.com/user-attachments/assets/9b28490f-f212-462e-814b-b8641cfe3c73" />

Look at the Clock visualizers. The top one is the clock of the digitalinput that I'm using to trigger ostim. The bottom one is the clock of when the ostim was delivered according to `Headstage64OpticalStimulatorData`. It seems the first trigger is ignored. The second trigger (labelled with green dot) applies the stimulus in a relatively timely manner (~1.5ms). The 3rd and 4th triggers (labelled with red dot) are super delayed (~13.8 seconds and9 seconds respectively) even though I'm sending a Double of value zero to the `Headstage64OpticalStimulatorTrigger` node.

This screenshot was made by taking the latest update to this branch and merging it to main.

jonnew committed:
Fixed ConfigureHeadstage64OpticalStimulator properties

- The pulse frequency was sometimes indicated as pulse period and other
  times as frequency with bad results

bparks13 committed:
Ensure consistent units between labels and scale bar

bparks13 committed:
Merge branch 'main' into issue-502-2

jonnew committed:
Match estim_device version 3 frame layout

jonnew committed:
Merge pull request #555 from open-ephys/issue-554

Match estim_device version 3 frame layout

jonnew reviewed (commented):
I wonder if the PortController changes should be cherry picked out of this PR even if the the `Headstage64GpoTrigger` and `ConfigureHeadstage64PortController` are removed. Why? Because having the ability to abstract the port controller's functions, especially control over GPIO, so that they appear to be properties of the headstage is a generally useful thing and this pattern allows that. I can foresee it being useful for other headtsages that have connections between the Serdes GPIO and the headtage. e.g. RHS headstage

EDIT: This was done in #569
